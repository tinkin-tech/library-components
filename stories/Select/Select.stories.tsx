import * as React from 'react'
import { SelectComponent, OptionsInterface } from './Select'

export default {
  title: 'Select',
  component: SelectComponent,
}

const options: OptionsInterface[] = [
  {
    id: 'option1',
    value: 'option1',
  },
  {
    id: 'option2',
    value: 'option2',
  },
  {
    id: 'option3',
    value: 'option3',
  },
]

interface PropsInterface {
  options: OptionsInterface[]
  className: string
  placeholder: string
  displayArrow: boolean
  error: string
  disabled: boolean
  search: boolean
  onChange: () => void
  id: string
  required?: boolean
  label?: string
}

const Template = (args: PropsInterface): React.ReactElement => (
  <SelectComponent {...args} />
)

export const Simple = Template.bind({})
Simple.args = {
  onChange: (): void => null,
  options: options || [],
  className: 'simple-select',
  placeholder: 'Select an option',
  displayArrow: true,
  error: '',
  disabled: '',
  search: false,
  id: 'select',
  label: 'Select Default',
  required: false,
}

export const Search = Template.bind({})
Search.args = {
  onChange: (): void => null,
  options: options || [],
  className: 'simple-select',
  placeholder: 'Select an option',
  displayArrow: true,
  error: '',
  disabled: '',
  search: true,
}
