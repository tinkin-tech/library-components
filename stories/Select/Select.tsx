import * as React from 'react'

export interface Option {
  id: string
  value: string
}

interface Props {
  onChange: (option: Option) => void
  options: Option[]
  displayArrow?: boolean
  className?: string
  valueId?: string
  placeholder?: string
  borderStyle?: boolean
  disable?: boolean
  error?: string
  search?: boolean
}

const SelectComponent = (props: Props): JSX.Element => {
  const {
    options,
    onChange,
    placeholder,
    valueId,
    className,
    displayArrow,
    borderStyle,
    disable,
    error,
    search,
  } = props

  const labels = {
    NO_OPTIONS: 'No hay opciones',
    SELECT: 'select',
  }

  const [expandedOptions, handleExpandedOptions] = React.useState<boolean>(
    false
  )

  const [selectedOption, handleSelectedOption] = React.useState<Option>({
    id: '',
    value: '',
  })

  const [shownOptions, handleShownOptions] = React.useState<Option[]>(options)

  const [inputValue, changeInputValue] = React.useState<string>(valueId || '')

  const selectOptions = React.useRef()

  const changeValue = (option: Option): void => {
    handleSelectedOption(option)
    handleExpandedOptions(false)
    onChange(option)
    changeInputValue(option.value)
  }

  const onSearchChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const newValue = event.currentTarget.value
    const newOptions = options.filter(
      (option) =>
        option.value.includes(newValue) || option.id.includes(newValue)
    )
    handleShownOptions(newOptions)
    handleSelectedOption(null)
    changeInputValue(newValue)
    handleExpandedOptions(true)
  }

  React.useEffect(() => {
    const eventListener = (event): void => {
      if (event.target !== selectOptions.current && expandedOptions) {
        handleExpandedOptions(false)
      }
    }
    // eslint-disable-next-line no-restricted-globals
    window.addEventListener('click', eventListener)
    // eslint-disable-next-line no-restricted-globals
    return (): void => window.removeEventListener('click', eventListener)
  })

  return (
    <div
      className={`select-component ${!disable && 'cursor-pointer'} ${
        disable && 'disabled'
      }`}
      onClick={(): void => !disable && handleExpandedOptions(!expandedOptions)}
      ref={selectOptions}
      data-testid="selectorComponent"
    >
      <div className="select-dropdown flex-1">
        <div
          className={`flex-row flex-middle flex-space-between flex-no-wrap ${
            borderStyle ? 'border radius-default bg-white border-padding' : ''
          } ${
            error
              ? 'border-warning'
              : expandedOptions
              ? 'border-primary'
              : 'border-secondary border-lighten-3'
          }`}
        >
          {search ? (
            <div className="input-container">
              <input
                value={inputValue}
                placeholder={placeholder}
                className={`flex-column p-r ${className || ''} ${
                  borderStyle ? '' : 'strong'
                } ${disable && 'disabled'}`}
                data-testid="selected-option"
                onChange={onSearchChange}
                onClick={(): void => handleExpandedOptions(!shownOptions)}
              />
            </div>
          ) : (
            <div
              className={`flex-column p-r ${className || ''} ${
                borderStyle ? '' : 'strong'
              } ${disable && 'disabled'}`}
              data-testid="selected-option"
            >
              {selectedOption.value || placeholder || ''}
            </div>
          )}

          {displayArrow && (
            <div className="flex-column p-r">
              <span className="arrow down" />
            </div>
          )}
        </div>
        {!disable && (
          <div
            className={`select-options app-scroll ${
              expandedOptions ? 'active' : ''
            } `}
          >
            {shownOptions.length > 0 ? (
              shownOptions.map((option: Option) => {
                return (
                  <span
                    key={option.id}
                    className="block option cursor-pointer"
                    data-testid={`${className}-${option.id}` || ''}
                    onClick={(): void => changeValue(option)}
                  >
                    {option.value}
                  </span>
                )
              })
            ) : (
              <span
                className="no-options block option cursor-default text-secondary text-lighten-3"
                onClick={(): null => null}
              >
                {labels.NO_OPTIONS}
              </span>
            )}
          </div>
        )}
      </div>
      {error && (
        <div className="flex-row flex-end size-small strong text-warning overflow-wrap">
          <div>{error}</div>
        </div>
      )}
    </div>
  )
}

export default React.memo(SelectComponent)
