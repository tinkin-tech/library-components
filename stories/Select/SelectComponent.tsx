import * as React from 'react'

import es_EC from './languages/es_EC'

interface IOption {
  id: string | number
  label: string
}

export interface ISelectComponent {
  options: IOption[]
  valueId: string
  onChangeValue: (id: string | number, valueId: string) => void
  value: string | number
  placeholder?: string
  label?: string
  error?: string
  required?: boolean
  readOnly?: boolean
  labelClassName?: string
  selectClassName?: string
  extraLabelClassName?: string
  extraSelectClassName?: string
}

export const SelectComponent: React.FC<ISelectComponent> = (
  props: ISelectComponent
) => {
  const {
    options,
    onChangeValue,
    valueId,
    value,
    placeholder,
    label,
    error,
    required,
    readOnly,
    labelClassName,
    selectClassName,
    extraLabelClassName,
    extraSelectClassName,
  } = props
  const [openSelector, handleOpenSelector] = React.useState(false)
  const selectRef = React.useRef(null)
  const onClickOptionItem = (itemId: string | number): void => {
    handleOpenSelector(false)
    onChangeValue(itemId, valueId)
  }
  const getValue =
    options.find((item) => item.id === value)?.label ||
    placeholder ||
    es_EC.placeholder

  const handleClickOutside = (event: Event): void => {
    if (!selectRef.current.contains(event.target)) {
      handleOpenSelector(false)
    }
  }

  React.useEffect(() => {
    document.addEventListener('click', handleClickOutside)
    return (): void => document.removeEventListener('click', handleClickOutside)
  })

  const selectClassNameObject = [
    selectClassName || 'select-component',
    extraSelectClassName || '',
    error ? 'select-component-error' : '',
    readOnly ? 'select-component-disable' : '',
  ].filter((item) => item !== '')

  const labelClassNameObject = [
    labelClassName || 'label',
    extraLabelClassName || '',
    error ? 'warning' : '',
    readOnly ? 'disable' : '',
  ].filter((item) => item !== '')

  return (
    <div className={selectClassNameObject.join(' ')} ref={selectRef}>
      {label && (
        <a
          className={labelClassNameObject.join(' ')}
          onClick={(): void => !readOnly && handleOpenSelector(true)}
        >
          {`${label}${required ? '*' : ''}`}
        </a>
      )}
      <a
        className="select-button"
        onClick={(): void => !readOnly && handleOpenSelector(true)}
      >
        {getValue}
        <i className="icon-arrow-down" />
      </a>
      {options.length && openSelector && (
        <ul className="selector-container">
          {options.map((item, index) => (
            <li
              className={`selector-item ${item.id === value ? 'selected' : ''}`}
              key={index}
            >
              <a onClick={(): void => onClickOptionItem(item.id)}>
                {item.label}
              </a>
            </li>
          ))}
        </ul>
      )}
      {error && <span className="error">{error}</span>}
    </div>
  )
}
