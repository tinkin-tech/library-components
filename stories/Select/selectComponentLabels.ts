export interface ISelectComponentLabels {
  NO_OPTIONS: string;
}

const selectComponentLabels: ISelectComponentLabels = {
  NO_OPTIONS: 'No hay opciones'
};

export default selectComponentLabels;
