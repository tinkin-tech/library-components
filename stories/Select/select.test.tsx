import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react'
import { Option } from './Select';
import Select from './Select';
import { action } from '@storybook/addon-actions'
import '@testing-library/jest-dom/extend-expect'
import addons, { mockChannel } from '@storybook/addons';

addons.setChannel(mockChannel());

const options: Option[] = [
  {
    id: 'option1',
    value: 'option1'
  },
  {
    id: 'option2',
    value: 'option2'
  },
  {
    id: 'option3',
    value: 'option3'
  }
]

const defaultSelect = (
  <Select
    onChange={action('selected')}
    options={options}
    className='simple-select'
    placeholder={'Select an option'}
    borderStyle={true}
    displayArrow={true}
    error={''}
    disable={false}
  />
)

describe('Select test', () => {

  it('Should render in virtual DOM', () => {
    const { getByTestId } = render(defaultSelect)
    expect(getByTestId('selectorComponent')).toBeInTheDocument()
  })

  it('Should shot the snapshot with all properties', () => {
    const component = renderer.create(defaultSelect)
    const componentJSON = component.toJSON()
    expect(componentJSON).toMatchSnapshot()
  })


  it('Should call action function one time when click option', async () => {
    const act = jest.fn();
    const compo = render(
      <Select
        onChange={act}
        options={options}
        className='simple-select'
        placeholder={'Select an option'}
        borderStyle={true}
        displayArrow={true}
        error={''}
        disable={false}
      />
    )
    renderer.act(() => {
      fireEvent.click(compo.getByTestId('simple-select-option1'));
    })
    expect(act).toBeCalledTimes(1);
  });

  it('should change component selected value', async () => {
    const act = jest.fn();
    const compo = render(
      <Select
        onChange={act}
        options={options}
        className='simple-select'
        placeholder={'Select an option'}
        borderStyle={true}
        displayArrow={true}
        error={''}
        disable={false}
      />
    )
    renderer.act(() => {
      fireEvent.click(compo.getByTestId('simple-select-option1'));
    })
    expect(compo.getByTestId('selected-option').innerHTML).toBe('option1');
  }) 
});
