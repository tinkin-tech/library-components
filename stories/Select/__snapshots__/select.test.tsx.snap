// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select test Should shot the snapshot with all properties 1`] = `
<div
  className="select-component cursor-pointer "
  data-testid="selectorComponent"
  onClick={[Function]}
>
  <div
    className="flex-space-between"
  >
    <label
      className="label "
      data-testid="label-component"
      htmlFor="select"
    >
      
      Â 
    </label>
    
  </div>
  <div
    className="select-dropdown flex-1"
  >
    <div
      className="flex-row flex-middle border radius-default flex-space-between flex-no-wrap border-gray border-lighten-1 bg-gray bg-lighten-3 "
    >
      <div
        className="input-container"
      >
        <input
          className="flex-column p-r simple-select false"
          data-testid="selected-option"
          disabled={true}
          onChange={[Function]}
          onClick={[Function]}
          placeholder="Select an option"
          value=""
        />
      </div>
      <div
        className="flex-column p-r arrow-dropdown"
      >
        <i
          className="arrow down icon-arrow"
        />
      </div>
    </div>
    <div
      className="select-options shadow-2 app-scroll  "
    >
      <span
        className="block option cursor-pointer"
        data-testid="simple-select-option1"
        onClick={[Function]}
      >
        option1
      </span>
      <span
        className="block option cursor-pointer"
        data-testid="simple-select-option2"
        onClick={[Function]}
      >
        option2
      </span>
      <span
        className="block option cursor-pointer"
        data-testid="simple-select-option3"
        onClick={[Function]}
      >
        option3
      </span>
    </div>
  </div>
  
</div>
`;
